aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/TokenProvidersDecorator.kt

```diff
@@ -0,0 +1,110 @@
+
+class TokenProvidersDecorator : ClientCodegenDecorator {
+    override val name: String get() = "TokenProvidersDecorator"
+    override val order: Byte = 0
+
+    private fun applies(codegenContext: ClientCodegenContext): Boolean =
+        ServiceIndex.of(codegenContext.model).getEffectiveAuthSchemes(codegenContext.serviceShape)
+            .containsKey(HttpBearerAuthTrait.ID)
+
+    override fun configCustomizations(
+        codegenContext: ClientCodegenContext,
+        baseCustomizations: List<ConfigCustomization>,
+    ): List<ConfigCustomization> =
+        if (applies(codegenContext)) {
+            baseCustomizations + TokenProviderConfig(codegenContext)
+        } else {
+            baseCustomizations
+        }
+
+    override fun extraSections(codegenContext: ClientCodegenContext): List<AdHocCustomization> =
+        if (applies(codegenContext)) {
+            listOf(
+                adhocCustomization<SdkConfigSection.CopySdkConfigToClientConfig> { section ->
+                    rust("${section.serviceConfigBuilder}.set_token_provider(${section.sdkConfig}.token_provider());")
+                },
+            )
+        } else {
+            emptyList()
+        }
+}
+
+/**
+ * Add a `.token_provider` field and builder to the `Config` for a given service
+ */
+class TokenProviderConfig(private val codegenContext: ClientCodegenContext) : ConfigCustomization() {
+    private val runtimeConfig = codegenContext.runtimeConfig
+    private val codegenScope =
+        arrayOf(
+            *RuntimeType.preludeScope,
+            "Token" to configReexport(AwsRuntimeType.awsCredentialTypes(runtimeConfig).resolve("Token")),
+            "ProvideToken" to
+                configReexport(
+                    AwsRuntimeType.awsCredentialTypes(runtimeConfig)
+                        .resolve("provider::token::ProvideToken"),
+                ),
+            "SharedTokenProvider" to
+                configReexport(
+                    AwsRuntimeType.awsCredentialTypes(runtimeConfig)
+                        .resolve("provider::token::SharedTokenProvider"),
+                ),
+            "TestToken" to AwsRuntimeType.awsCredentialTypesTestUtil(runtimeConfig).resolve("Token"),
+            "HTTP_BEARER_AUTH_SCHEME_ID" to
+                CargoDependency.smithyRuntimeApiClient(runtimeConfig)
+                    .withFeature("http-auth").toType().resolve("client::auth::http").resolve("HTTP_BEARER_AUTH_SCHEME_ID"),
+        )
+
+    override fun section(section: ServiceConfig) =
+        writable {
+            when (section) {
+                ServiceConfig.BuilderImpl -> {
+                    rustTemplate(
+                        """
+                        /// Sets the access token provider for this service
+                        pub fn token_provider(mut self, token_provider: impl #{ProvideToken} + 'static) -> Self {
+                            self.set_token_provider(#{Some}(#{SharedTokenProvider}::new(token_provider)));
```

Comment:
should this utilize `IntoShared`?

Reply:
Done in e7bfbce9e8e3881c18f10e57ebadb119af31aa2c